File: PagueMais/Purchase/entities/Purchase.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

public enum EnumStatus
{
    Pending,
    Completed,
    Canceled
}

public enum EnumMethods
{
    Credit,
    Debit,
    Pix,
    Money
}

namespace Entities
{
    public class Purchase
    {
        public Purchase()
        {

        }

        public Purchase(float Total, EnumMethods PaymentMethod, Guid ClientId)
        {
            this.Total = Total;
            this.PaymentMethod = PaymentMethod;
            this.ClientId = ClientId;
        }

        [Key]
        public Guid Id { get; set; } = new();
        public float Total { get; set; }
        public EnumMethods? PaymentMethod { get; set; }
        public EnumStatus Status { get; set; } = EnumStatus.Pending;
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? FinishedAt { get; set; }
        public DateTime? UpdatedAt { get; set; } = null;

        [ForeignKey("Client")]
        public Guid ClientId { get; set; }
        public Client? Client { get; set; }

        public IEnumerable<Cart>? Carts { get; set; }
    }
}

File: client/src/types/purchase.ts
import { Client } from "./client";
import { Product } from "./product";
import { DialogUIStates } from "./ui";

enum PaymentMethods {
  Credit,
  Debit,
  Pix,
  Money,
}

enum Status {
  Pending,
  Completed,
  Canceled,
}

export type Purchase = {
  id: string;
  total: number;
  paymentMethod: PaymentMethods;
  status: Status;
  clientId: string;
  createdAt: Date;
  finishedAt: Date;
  updatedAt?: Date;
  carts?: Cart[];
};

export interface CreatePurchaseSchema {
  paymentMethod: PaymentMethods;
  clientId: string;
}

export interface UpdatePurchaseSchema {
  paymentMethod: PaymentMethods;
  status: Status;
}

export type Cart = {
  id: string;
  quantity: number;
  productId: string;
  product: Product;
  purchaseId: string;
  purchase: Purchase;
  createdAt: Date;
  updatedAt?: Date;
};

export interface CreateCartSchema {
  productId: string;
  purchaseId: string;
}

export type PurchaseDialogUIStates = DialogUIStates | "addingProducts";

export interface UsePurchaseUIState {
  dialogUIState: PurchaseDialogUIStates;
  setDialogUIState: (UIState: PurchaseDialogUIStates) => void;
}

export interface UsePurchaseState {
  purchase: Purchase | null;
  product: Product | null;
  clientId: string | null;
  setPurchase: (purchase: Purchase | null) => void;
  setProduct: (product: Product | null) => void;
  setClientId: (client: string | null) => void;
  closeDialog: () => void;
  setIsDeleting: (purchase: Purchase) => void;
  setIsUpdating: (purchase: Purchase) => void;
  setIsAddingProducts: () => void;
}
